File: src/ipc/mock/handlers/branches.ts
// Split out of the former monolithic mock.ts (pure refactor; no behavior change).
import type { IpcApi } from '../../types';
import { randomOid } from '../../fixtures/oids';
import { buildStaleReport, delay, isInvalidBranchName, query, requireRepo } from '../repoState';
import { upsert } from '../statusHelpers';
import type { AppError, BranchDeleteResult, BranchDeleteStatus, BranchesSnapshot, CheckoutResult, CreateBranchHereResult, StaleReport } from '../../types';

export const branchHandlers = {
  async listBranches(repoId: string): Promise<BranchesSnapshot> {
    await delay(150);
    const state = requireRepo(repoId);
    const snapshot = structuredClone(state.branches);
    if (state.kind === 'detached') {
      snapshot.head = { branchName: null, oid: state.headOid, detached: true, unborn: false };
      for (const branch of snapshot.local) branch.isHead = false;
    } else {
      snapshot.head = {
        branchName: state.headBranch,
        oid: state.headOid,
        detached: false,
        unborn: false,
      };
    }
    return snapshot;
  },

  async createBranch(repoId: string, name: string): Promise<void> {
    await delay(150);
    const state = requireRepo(repoId);
    if (isInvalidBranchName(name)) {
      const err: AppError = { kind: 'invalidName', message: `invalid branch name: '${name}'` };
      throw err;
    }
    const trimmed = name.trim();
    if (state.branches.local.some((b) => b.name === trimmed)) {
      const err: AppError = {
        kind: 'branchExists',
        message: `branch '${trimmed}' already exists`,
      };
      throw err;
    }
    state.branches.local.push({
      name: trimmed,
      isHead: false,
      upstream: null,
      ahead: null,
      behind: null,
      tip: randomOid(),
    });
    state.branches.local.sort((a, b) => a.name.toLowerCase().localeCompare(b.name.toLowerCase()));
  },

  // P11 §1.3: create a local branch at `oid` and check it out, carrying any
  // dirty worktree across. Stateful so the graph HEAD/new-branch pills move on
  // the next refreshAll. `?branch=cbhconflict` exercises the Conflicts toast.
  async createBranchHere(
    repoId: string,
    name: string,
    oid: string,
  ): Promise<CreateBranchHereResult> {
    await delay(250);
    const state = requireRepo(repoId);
    if (isInvalidBranchName(name)) {
      const err: AppError = { kind: 'invalidName', message: `invalid branch name: '${name}'` };
      throw err;
    }
    const trimmed = name.trim();
    if (state.branches.local.some((b) => b.name === trimmed)) {
      const err: AppError = {
        kind: 'branchExists',
        message: `branch '${trimmed}' already exists`,
      };
      throw err;
    }
    const s = state.status;
    const dirty =
      s.staged.length > 0 ||
      s.unstaged.length > 0 ||
      s.untracked.length > 0 ||
      s.conflicted.length > 0;
    // Add the new branch at `oid` as the checked-out HEAD (unset previous head)
    // + move headBranch/headOid so the graph HEAD pill follows on refreshAll.
    for (const b of state.branches.local) b.isHead = false;
    state.branches.local.push({
      name: trimmed,
      isHead: true,
      upstream: null,
      ahead: null,
      behind: null,
      tip: oid,
    });
    state.branches.local.sort((a, b) => a.name.toLowerCase().localeCompare(b.name.toLowerCase()));
    state.headBranch = trimmed;
    state.headOid = oid;
    state.branches.head = { branchName: trimmed, oid, detached: false, unborn: false };
    if (!dirty) return { stashed: false, apply: null };
    // Simulate carrying work across the switch.
    if (query('branch') === 'cbhconflict') {
      // Carried with markers: worktree stays dirty (synthetic conflict entry)
      // and the stash would be RETAINED — do NOT clear the status.
      upsert(s.conflicted, { path: 'src/app.ts', origPath: null, status: 'conflicted' });
      return { stashed: true, apply: { kind: 'conflicts', paths: ['src/app.ts'] } };
    }
    // Clean carry-over: the changes moved with us — status preserved as-is.
    return { stashed: true, apply: { kind: 'applied' } };
  },

  // P33: dirty-safe switch — auto-stash → switch → auto fast-forward (no fetch)
  // → re-apply stash. Never hard-fails on a dirty tree; a conflicted re-apply is
  // a SUCCESS carrying `apply: {kind:'conflicts'}` (stash RETAINED).
  async checkoutBranch(repoId: string, name: string): Promise<CheckoutResult> {
    await delay(150);
    const state = requireRepo(repoId);
    // P36: deterministic worktree-collision refusal — a reserved fixture branch
    // name simulates the branch being checked out in another worktree.
    if (name === '__wt_locked__') {
      const err: AppError = {
        kind: 'branchCheckedOutElsewhere',
        message: `branch '${name}' is already checked out at '/repo/.worktrees/${name}'`,
      };
      throw err;
    }
    const branch = state.branches.local.find((b) => b.name === name);
    if (branch === undefined) {
      const err: AppError = { kind: 'branchNotFound', message: `branch '${name}' not found` };
      throw err;
    }
    const s = state.status;
    const dirty =
      s.staged.length > 0 ||
      s.unstaged.length > 0 ||
      s.untracked.length > 0 ||
      s.conflicted.length > 0;

    // Move HEAD to the target branch (unset previous head).
    for (const b of state.branches.local) b.isHead = false;
    branch.isHead = true;
    state.headBranch = name;
    state.branches.head = { branchName: name, oid: state.headOid, detached: false, unborn: false };
    // TODO(polish): move the HEAD/branch pills in the mock graph fixture too
    // (contract §5 decision: fixtures stay decoupled from branch state —
    // harness proof is the sidebar dot + header branch name).

    // Auto fast-forward: only when the target tracks an upstream and is strictly
    // behind (behind>0 && ahead==0). `feature/merged-a` is the deterministic FF
    // fixture (ahead 0, behind 3). Diverged (feature/sidebar) or up-to-date
    // (main) → no FF.
    const fastForwarded =
      branch.upstream != null && (branch.behind ?? 0) > 0 && (branch.ahead ?? 0) === 0;
    if (fastForwarded) {
      branch.behind = 0;
    }

    if (!dirty) return { stashed: false, fastForwarded, apply: null };

    // Carried work across the switch. `fix/watcher-debounce` is the designated
    // conflicted re-apply fixture (contract §4.3): worktree stays dirty (stash
    // RETAINED) with a synthetic conflict entry.
    if (name === 'fix/watcher-debounce') {
      upsert(s.conflicted, { path: 'src/app.ts', origPath: null, status: 'conflicted' });
      return { stashed: true, fastForwarded, apply: { kind: 'conflicts', paths: ['src/app.ts'] } };
    }
    // Clean carry-over: the changes moved with us — status preserved as-is.
    return { stashed: true, fastForwarded, apply: { kind: 'applied' } };
  },

  async deleteBranch(repoId: string, name: string): Promise<void> {
    await delay(150);
    const state = requireRepo(repoId);
    const branch = state.branches.local.find((b) => b.name === name);
    if (branch === undefined) {
      const err: AppError = { kind: 'branchNotFound', message: `branch '${name}' not found` };
      throw err;
    }
    if (branch.isHead) {
      const err: AppError = {
        kind: 'git',
        message: `cannot delete '${name}': it is the currently checked-out branch`,
      };
      throw err;
    }
    // Designated unmerged branch (contract §5).
    if (name === 'experiment-unmerged') {
      const err: AppError = {
        kind: 'unmergedBranch',
        message:
          "branch 'experiment-unmerged' is not fully merged into HEAD (tip 1a2b3c4). " +
          'Bonsai v1 does not force-delete; use `git branch -D experiment-unmerged` ' +
          'if you are sure.',
      };
      throw err;
    }
    state.branches.local = state.branches.local.filter((b) => b.name !== name);
  },

  // P6 §3.5: GitKraken-style remote checkout — create/reuse a local tracking
  // branch for the remote-tracking ref and switch to it.
  async checkoutRemoteBranch(repoId: string, name: string): Promise<void> {
    await delay(150);
    const state = requireRepo(repoId);
    const remote = state.branches.remote.find((r) => r.name === name);
    if (remote === undefined) {
      const err: AppError = {
        kind: 'branchNotFound',
        message: `remote-tracking branch '${name}' not found`,
      };
      throw err;
    }
    // Split on the FIRST '/' (remote names contain no '/').
    const slash = name.indexOf('/');
    const localName = slash === -1 ? name : name.slice(slash + 1);
    let local = state.branches.local.find((b) => b.name === localName);
    if (local === undefined) {
      // Create-and-track path: new local tracking branch at the remote tip.
      local = { name: localName, isHead: false, upstream: name, ahead: 0, behind: 0, tip: remote.tip };
      state.branches.local.push(local);
      state.branches.local.sort((a, b) => a.name.toLowerCase().localeCompare(b.name.toLowerCase()));
    }
    // Switch HEAD (same state transition as checkoutBranch).
    for (const b of state.branches.local) b.isHead = false;
    local.isHead = true;
    state.headBranch = local.name;
    state.headOid = local.tip;
    state.branches.head = {
      branchName: local.name,
      oid: state.headOid,
      detached: false,
      unborn: false,
    };
  },

  // P6 §3.5: delete the LOCAL remote-tracking ref only (never touches the server).
  async deleteRemoteBranch(repoId: string, name: string): Promise<void> {
    await delay(150);
    const state = requireRepo(repoId);
    const remote = state.branches.remote.find((r) => r.name === name);
    if (remote === undefined) {
      const err: AppError = {
        kind: 'branchNotFound',
        message: `remote-tracking branch '${name}' not found`,
      };
      throw err;
    }
    state.branches.remote = state.branches.remote.filter((r) => r.name !== name);
  },

  // P25 §6.3: stale-branch cleanup. `listStaleBranches` recomputes the report
  // from the live branch list; `deleteBranches` mirrors the server's re-verified
  // safety rules and MUTATES `state.branches.local` for every deleted name so the
  // harness shows rows disappear + a summary toast.
  async listStaleBranches(repoId: string, _base?: string): Promise<StaleReport> {
    await delay(150);
    const state = requireRepo(repoId);
    return buildStaleReport(state);
  },

  async deleteBranches(
    repoId: string,
    names: string[],
    _base?: string,
  ): Promise<BranchDeleteResult[]> {
    await delay(200);
    const state = requireRepo(repoId);
    const report = buildStaleReport(state);
    const safe = new Set(report.branches.map((b) => b.name));
    const currentName = state.kind === 'detached' ? null : state.headBranch;

    const results: BranchDeleteResult[] = names.map((name) => {
      if (name === currentName) {
        return { name, status: 'skippedCurrent' as BranchDeleteStatus, message: 'checked-out branch' };
      }
      if (name === report.base) {
        return { name, status: 'skippedBase' as BranchDeleteStatus, message: 'base branch' };
      }
      if (!safe.has(name)) {
        return {
          name,
          status: 'skippedNotStale' as BranchDeleteStatus,
          message: 'not detected as stale',
        };
      }
      // Safe → remove it from the live branch list (mutating shrink).
      state.branches.local = state.branches.local.filter((b) => b.name !== name);
      return { name, status: 'deleted' as BranchDeleteStatus, message: null };
    });
    return results;
  },

  // Stateful remote mock (M6 contract §5). Failure triggers via `?remote=`
  // (authfail | network | rejected | conflict), composable with `?fixture=`.
} satisfies Partial<IpcApi>;

File: src-tauri/src/commands/branches.rs
//! `branches` commands — split from the former monolithic `commands.rs`.

use super::shared::*;

/// One snapshot of local branches + remote-tracking branches + tags + HEAD
/// (M5 contract §2.2/§2.8). Errors: `noRepo` | `git`.
#[tauri::command]
pub async fn list_branches(
    state: tauri::State<'_, AppState>,
    repo_id: String,
) -> Result<BranchesSnapshot, AppError> {
    list_branches_inner(state.inner(), &repo_id).await
}

/// Runtime-free core of `list_branches` (unit-testable without a Tauri app).
pub(crate) async fn list_branches_inner(
    state: &AppState,
    repo_id: &str,
) -> Result<BranchesSnapshot, AppError> {
    // P88b/B2b: route through the per-thread handle cache. Refs/tags/HEAD are
    // re-read on demand, so a reused handle reads current on-disk state.
    let (path, generation) = repo_path_and_gen(state, repo_id)?;
    let perf = state.perf.clone();
    let repo_id = repo_id.to_string();
    tauri::async_runtime::spawn_blocking(move || {
        crate::repo_handle::with_repo(&repo_id, generation, &path, &perf, |repo| {
            branches::list_refs_with(repo)
        })
    })
    .await
    .map_err(|e| AppError::Other(format!("task join error: {e}")))?
}

/// Creates a local branch at the current HEAD commit — does NOT check out
/// (M5 contract §2.4). Errors: `invalidName` | `branchExists` | `git` | `noRepo`.
/// Does NOT emit `repo-changed` — the frontend refetches imperatively.
#[tauri::command]
pub async fn create_branch(
    state: tauri::State<'_, AppState>,
    repo_id: String,
    name: String,
) -> Result<(), AppError> {
    create_branch_inner(state.inner(), &repo_id, name).await
}

/// Runtime-free core of `create_branch` (unit-testable without a Tauri app).
pub(crate) async fn create_branch_inner(
    state: &AppState,
    repo_id: &str,
    name: String,
) -> Result<(), AppError> {
    let path = repo_path(state, repo_id)?;
    tauri::async_runtime::spawn_blocking(move || branches::create_branch(&path, &name))
        .await
        .map_err(|e| AppError::Other(format!("task join error: {e}")))?
}

/// Creates local branch `name` at commit `oid`, auto-stashing/re-applying
/// uncommitted work across the checkout (P11 §1). Errors: `invalidName` |
/// `branchExists` | `operationInProgress` | `configMissing` | `checkoutConflict`
/// | `git` | `noRepo`. Does NOT emit `repo-changed`.
#[tauri::command]
pub async fn create_branch_here(
    state: tauri::State<'_, AppState>,
    repo_id: String,
    name: String,
    oid: String,
) -> Result<CreateBranchHereResult, AppError> {
    create_branch_here_inner(state.inner(), &repo_id, name, oid).await
}

/// Runtime-free core of `create_branch_here` (unit-testable without a Tauri app).
pub(crate) async fn create_branch_here_inner(
    state: &AppState,
    repo_id: &str,
    name: String,
    oid: String,
) -> Result<CreateBranchHereResult, AppError> {
    let path = repo_path(state, repo_id)?;
    tauri::async_runtime::spawn_blocking(move || branches::create_branch_here(&path, &name, &oid))
        .await
        .map_err(|e| AppError::Other(format!("task join error: {e}")))?
}

/// Dirty-safe checkout of a LOCAL branch (P33): auto-stash -> switch -> auto FF
/// to upstream (no fetch) -> re-apply stash. A conflicted re-apply is a SUCCESS
/// carrying `apply: Some(conflicts)` (stash retained). Errors: `branchNotFound`
/// | `operationInProgress` | `configMissing` | `checkoutConflict` | `git` |
/// `noRepo`. Does NOT emit `repo-changed` (frontend calls refreshAll).
#[tauri::command]
pub async fn checkout_branch(
    state: tauri::State<'_, AppState>,
    repo_id: String,
    name: String,
) -> Result<CheckoutResult, AppError> {
    checkout_branch_inner(state.inner(), &repo_id, name).await
}

/// Runtime-free core of `checkout_branch` (unit-testable without a Tauri app).
pub(crate) async fn checkout_branch_inner(
    state: &AppState,
    repo_id: &str,
    name: String,
) -> Result<CheckoutResult, AppError> {
    let path = repo_path(state, repo_id)?;
    tauri::async_runtime::spawn_blocking(move || branches::checkout_branch_autostash(&path, &name))
        .await
        .map_err(|e| AppError::Other(format!("task join error: {e}")))?
}

/// Dirty-safe checkout of an arbitrary commit → DETACHED HEAD: auto-stash ->
/// safe checkout -> set_head_detached -> re-apply stash. No auto-FF. A conflicted
/// re-apply is a SUCCESS carrying `apply: Some(conflicts)` (stash retained).
/// Errors: `invalidName` | `git` | `operationInProgress` | `configMissing` |
/// `checkoutConflict` | `noRepo`. Does NOT emit `repo-changed` (frontend calls
/// refreshAll).
#[tauri::command]
pub async fn checkout_commit(
    state: tauri::State<'_, AppState>,
    repo_id: String,
    oid: String,
) -> Result<CheckoutResult, AppError> {
    checkout_commit_inner(state.inner(), &repo_id, oid).await
}

/// Runtime-free core of `checkout_commit` (unit-testable without a Tauri app).
pub(crate) async fn checkout_commit_inner(
    state: &AppState,
    repo_id: &str,
    oid: String,
) -> Result<CheckoutResult, AppError> {
    let path = repo_path(state, repo_id)?;
    tauri::async_runtime::spawn_blocking(move || branches::checkout_commit_detached(&path, &oid))
        .await
        .map_err(|e| AppError::Other(format!("task join error: {e}")))?
}

/// Deletes a LOCAL, fully merged, non-current branch (M5 contract §2.6 —
/// unmerged deletion is blocked; no force-delete in v1).
/// Errors: `branchNotFound` | `unmergedBranch` | `git` | `noRepo`.
#[tauri::command]
pub async fn delete_branch(
    state: tauri::State<'_, AppState>,
    repo_id: String,
    name: String,
) -> Result<(), AppError> {
    delete_branch_inner(state.inner(), &repo_id, name).await
}

/// Runtime-free core of `delete_branch` (unit-testable without a Tauri app).
pub(crate) async fn delete_branch_inner(
    state: &AppState,
    repo_id: &str,
    name: String,
) -> Result<(), AppError> {
    let path = repo_path(state, repo_id)?;
    tauri::async_runtime::spawn_blocking(move || branches::delete_branch(&path, &name))
        .await
        .map_err(|e| AppError::Other(format!("task join error: {e}")))?
}

/// Renames LOCAL branch `old_name` → `new_name` (git `branch -m`, non-force,
/// P60a). Preserves upstream + reflog; rewrites HEAD when `old_name` is the
/// checked-out branch. Errors: `invalidName` | `branchNotFound` | `branchExists`
/// | `git` | `noRepo`. Does NOT emit `repo-changed` — the frontend refetches.
#[tauri::command]
pub async fn rename_branch(
    state: tauri::State<'_, AppState>,
    repo_id: String,
    old_name: String,
    new_name: String,
) -> Result<RenameBranchResult, AppError> {
    rename_branch_inner(state.inner(), &repo_id, old_name, new_name).await
}

/// Runtime-free core of `rename_branch` (unit-testable without a Tauri app).
pub(crate) async fn rename_branch_inner(
    state: &AppState,
    repo_id: &str,
    old_name: String,
    new_name: String,
) -> Result<RenameBranchResult, AppError> {
    let path = repo_path(state, repo_id)?;
    tauri::async_runtime::spawn_blocking(move || {
        branches::rename_branch(&path, &old_name, &new_name)
    })
    .await
    .map_err(|e| AppError::Other(format!("task join error: {e}")))?
}

/// GitKraken-style remote checkout: create/reuse a local tracking branch for
/// `name` ("<remote>/<branch>") and safe-checkout it (P6 §2.2).
/// Errors: `invalidName` | `branchNotFound` | `checkoutConflict` | `git` | `noRepo`.
#[tauri::command]
pub async fn checkout_remote(
    state: tauri::State<'_, AppState>,
    repo_id: String,
    name: String,
) -> Result<(), AppError> {
    checkout_remote_inner(state.inner(), &repo_id, name).await
}

/// Runtime-free core of `checkout_remote` (unit-testable without a Tauri app).
pub(crate) async fn checkout_remote_inner(
    state: &AppState,
    repo_id: &str,
    name: String,
) -> Result<(), AppError> {
    let path = repo_path(state, repo_id)?;
    tauri::async_runtime::spawn_blocking(move || branches::checkout_remote(&path, &name))
        .await
        .map_err(|e| AppError::Other(format!("task join error: {e}")))?
}

/// Deletes the LOCAL remote-tracking ref `name` — does NOT touch the server
/// (P6 §2.3). Errors: `branchNotFound` | `git` | `noRepo`.
#[tauri::command]
pub async fn delete_remote_tracking(
    state: tauri::State<'_, AppState>,
    repo_id: String,
    name: String,
) -> Result<(), AppError> {
    delete_remote_tracking_inner(state.inner(), &repo_id, name).await
}

/// Runtime-free core of `delete_remote_tracking` (unit-testable without a Tauri app).
pub(crate) async fn delete_remote_tracking_inner(
    state: &AppState,
    repo_id: &str,
    name: String,
) -> Result<(), AppError> {
    let path = repo_path(state, repo_id)?;
    tauri::async_runtime::spawn_blocking(move || branches::delete_remote_tracking(&path, &name))
        .await
        .map_err(|e| AppError::Other(format!("task join error: {e}")))?
}

/// Classifies local branches safe to delete (merged into `base` OR
/// upstream-gone) — read-only, touches nothing (P25 §4.1). `base` auto-resolves
/// when omitted. Pure git; NO consent gate. Errors: `git` | `noRepo`.
#[tauri::command]
pub async fn list_stale_branches(
    state: tauri::State<'_, AppState>,
    repo_id: String,
    base: Option<String>,
) -> Result<StaleReport, AppError> {
    list_stale_branches_inner(state.inner(), &repo_id, base).await
}

/// Runtime-free core of `list_stale_branches` (unit-testable without a Tauri app).
pub(crate) async fn list_stale_branches_inner(
    state: &AppState,
    repo_id: &str,
    base: Option<String>,
) -> Result<StaleReport, AppError> {
    let path = repo_path(state, repo_id)?;
    tauri::async_runtime::spawn_blocking(move || {
        stale::find_stale_branches(&path, base.as_deref())
    })
    .await
    .map_err(|e| AppError::Other(format!("task join error: {e}")))?
}

/// Batch-deletes the caller-supplied branch names that are STILL safe against a
/// freshly-recomputed stale set — refusing the current branch, the base, and
/// anything not re-verified as stale (P25 §4.3). Per-branch outcomes are DATA,
/// never thrown; a partial batch returns `Ok(results)`. Pure git; NO consent
/// gate. Does NOT emit `repo-changed` — the frontend refetches imperatively.
/// Errors (whole-call): `git` (bad base) | `noRepo`.
#[tauri::command]
pub async fn delete_branches(
    state: tauri::State<'_, AppState>,
    repo_id: String,
    names: Vec<String>,
    base: Option<String>,
) -> Result<Vec<BranchDeleteResult>, AppError> {
    delete_branches_inner(state.inner(), &repo_id, names, base).await
}

/// Runtime-free core of `delete_branches` (unit-testable without a Tauri app).
pub(crate) async fn delete_branches_inner(
    state: &AppState,
    repo_id: &str,
    names: Vec<String>,
    base: Option<String>,
) -> Result<Vec<BranchDeleteResult>, AppError> {
    let path = repo_path(state, repo_id)?;
    tauri::async_runtime::spawn_blocking(move || {
        stale::delete_branches(&path, &names, base.as_deref())
    })
    .await
    .map_err(|e| AppError::Other(format!("task join error: {e}")))?
}
